src/ledger/shred_inserter/merkle_root_checks.zig

```diff
@@ -228,43 +197,88 @@ pub fn checkBackwardsChainedMerkleRootConsistency(
         .index = @intCast(prev_erasure_meta.first_received_code_index),
         .shred_type = .code,
     };
-    const prev_shred =
-        if (try shred_store.get(prev_shred_id)) |ps| ps else {
-        logger.warn().logf(&newlinesToSpaces(
-            \\Shred {any} indicated by the erasure meta {any} \
-            \\is missing from blockstore. This can happen if you have recently upgraded \
-            \\from a version < v1.18.13, or if blockstore cleanup has caught up to the root. \
-            \\Skipping the backwards chained merkle root consistency check
-        ), .{ prev_shred_id, prev_erasure_meta });
+
+    return checkAndReportMerkleRootConsistency(
+        .backward,
+        logger,
+        db,
+        shred_store,
+        duplicate_shreds,
+        shred,
+        prev_shred_id,
+    );
+}
+
+/// The input shreds must be from adjacent erasure sets in the same slot,
+/// or this function will not work correctly.
+fn checkAndReportMerkleRootConsistency(
+    direction: enum { forward, backward },
+    logger: Logger,
+    db: *BlockstoreDB,
+    shred_store: WorkingShredStore,
+    duplicate_shreds: *std.ArrayList(PossibleDuplicateShred),
+    shred: Shred,
+    other_shred_id: ShredId,
+) !bool {
+    const other_shred = if (try shred_store.get(other_shred_id)) |other_shred|
+        other_shred
+    else {
+        logger.warn().logf(
+            "Shred {any} is missing from blockstore. " ++
+                "This can happen if blockstore cleanup has caught up to the root. " ++
+                "Skipping the {} chained merkle root consistency check.",
+            .{ other_shred_id, direction },
+        );
         return true;
     };
-    const merkle_root = shred_mod.layout.getChainedMerkleRoot(prev_shred);
-    const chained_merkle_root = shred.chainedMerkleRoot() catch null;
 
-    if (!checkChaining(merkle_root, chained_merkle_root)) {
-        logger.warn().logf(&newlinesToSpaces(
-            \\Received conflicting chained merkle roots for slot: {}, shred {any} type {any} \
-            \\chains to merkle root {any}, however previous fec set code \
-            \\shred {any} has merkle root {any}. Reporting as duplicate
-        ), .{
+    const early_shred, const late_shred = switch (direction) {
+        .forward => .{ shred.payload(), other_shred },
+        .backward => .{ other_shred, shred.payload() },
+    };
+
+    const chained_merkle_root = shred_mod.layout.getChainedMerkleRoot(late_shred);
```

Comment:
Agave has this as

```
let merkle_root = shred::layout::get_merkle_root(&prev_shred);
 let chained_merkle_root = shred.chained_merkle_root().ok();
```

where 
 - `prev_shred` -> `merkle_root` and
 - `current_shred` -> `chained_merkle_root`

You seem to have flipped that into

- `late_shred` aka `prev_shred` -> `chained_merkle_root`
- `early_shred` aka current shred - > `merkle_root`

and I guess that is because it is only the later shred that can only be considered to have chained merkle root? and not the one currently being processed?

Reply:
> late_shred aka prev_shred -> chained_merkle_root
> early_shred aka current shred - > merkle_root

You have it backwards. `prev_shred` in the backward chaining function becomes `early_shred` in this function.

Late shred comes after early shred. A shred's chained merkle root is always supposed to point to the erasure set that came before it. So the late shred's chained merkle root is supposed to equal the early shred's merkle root.